src/main/java/com/nirima/jenkins/plugins/docker/DockerTemplate.java

```diff
@@ -412,35 +409,25 @@ public void onNext(PullResponseItem item) {
 
     }
 
+    public final CreateContainerCmd createContainerCmd(DockerAPI api){
+        final DockerClient client = api.getClient();
+        CreateContainerCmd cmd = client.createContainerCmd(this.getImage());
+        this.fillContainerConfig(cmd);
+        return cmd;
+    }
+
     @Restricted(NoExternalUse.class)
     public DockerTransientNode provisionNode(DockerAPI api, TaskListener listener) throws IOException, Descriptor.FormException, InterruptedException {
 
-        final DockerClient client = api.getClient();
         final DockerComputerConnector connector = getConnector();
         pullImage(api, listener);
 
-        LOGGER.info("Trying to run container for {}", getImage());
-        CreateContainerCmd cmd = client.createContainerCmd(getImage());
-        fillContainerConfig(cmd);
-
-        connector.beforeContainerCreated(api, remoteFs, cmd);
-
-        String containerId = cmd.exec().getId();
-
-        try {
-            connector.beforeContainerStarted(api, remoteFs, containerId);
-
-            client.startContainerCmd(containerId).exec();
-
-            connector.afterContainerStarted(api, remoteFs, containerId);
-        } catch (DockerException e) {
-            // if something went wrong, cleanup aborted container
-            client.removeContainerCmd(containerId).withForce(true).exec();
-            throw e;
-        }
```

Comment:
You don't need to move this code away and introduce `DockerContainerLifecycleHandler`(which makes DockerComputerConnector role unclear). Just compute `containerName`, create `DockerTransientNode ` object, then enforce here `CreateContainerCmd.name` is set to this unique ID and pass node object to connector methods for launcher setup.

Reply:
I moved this code to the DockerComputerConnector, since it must be called from the launcher itself and cannot run before the launcher is being created. In this iteration, I tried to keep the original code as it was when I could (sorry for the many refactoring - did not try to make your review hard, it was just multiple iterations).
Also removed the DockerContainerLifecycleHandler and restored the original methods inside the DockerComputerConnector)